resources/knowledgebase/tla-review-guidelines.md

```diff
@@ -0,0 +1,147 @@
+  ```
+  Here, `waiting` and `id` are redundantly restated. Use nested `EXCEPT` syntax instead to specify only what changes:
+  ```tla
+  producers' = [producers EXCEPT ![p].count = @ + 1]
+  ```
+  See `tla-functions-records-sequences.md` for details on `EXCEPT` and nested updates.
+
+## Spec Structure
+
+* Use SANY to perform syntax and level-checking of a TLA+ module before reviewing. This catches syntax errors early and ensures the spec is well-formed. If you are an AI assistant, use the `tlaplus_mcp_sany_parse` tool to parse the spec.
+
+### Module
+
+* The TLA+ CommunityModules are a collection of modules that provide useful operators and definitions for TLA+. Use of the CommunityModules is encouraged because most are well-tested and well-documented. Some operators even have TLC Java overrides that make them more efficient. The CommunityModules are available at https://github.com/tlaplus/CommunityModules, but they are already readily available in the VSCode/Cursor extension without manual installation.
+
+* Use the `tlaplus_mcp_sany_modules` tool to list all available TLA+ modules, including the CommunityModules. Use the `tlaplus_mcp_sany_symbol` tool to extract and inspect the symbols (constants, operators, definitions) defined in any module.
+
+* `EXTENDS TLC` is rarely needed in ordinary TLA+ specs. The operators defined in the TLC module are for special use cases.
+  - The only notable exception is the `TLC!@@` operator, which merges two functions.
+  - The use of `RandomElement` is almost always wrong. Use non-determinism (`\E` or disjunction) instead. See `tla-RandomElement.md` for details.
+  - Use of `TLC!Assert` is discouraged in plain TLA+ specs. `Assert` is primarily used in PlusCal specs. An invariant is a better alternative, and TLC will report a counterexample if the invariant is violated.
+
+### Constants and Assumptions
+
+* If the spec declares constants, state assumptions that constrain their values. Ensure all constants have appropriate assumptions - for example, if declaring constants `NumProducers` and `NumConsumers`, use `ASSUME NumProducers \in Nat \ {0}` and `ASSUME NumConsumers \in Nat \ {0}` rather than only constraining one. Assumptions should be comprehensive and prevent nonsensical constant combinations (e.g., ensuring `BufferSize > 0`). TLC will refuse to check the spec if assumptions are violated, and assumptions aid readers in understanding the spec and are crucial for proving properties with TLAPS. Note: assumptions can be combined into a single ASSUME statement using conjunction, but TLC will not report which conjunct is violated, only that the assumption is violated.
+
+* Prefer using model values over integer intervals for constants representing actors or entities in the system. For example, instead of defining `Producers == 1..NumProducers`, make `Producers` a CONSTANT whose value is defined in the TLC configuration like `{p1, p2}`. This improves readability and makes traces easier to understand.
```

Comment:
I'd rephrase that a little. Sets of model values for constant parameters are particularly useful when that parameter is a symmetry set. Sometimes order is important (for example for representing nodes arranged on a ring), and then it is perfectly OK to use integer intervals.